File: java/is_valid_socket_address/Answer.java
import java.util.regex.Matcher;
import java.util.regex.Pattern;
/*
 * This is the file with your answer, do not rename or move it.
 * Write your code in it, and save it before submitting your answer.
 */

/*
* Write the code for your answer in this class.
*/
public class Answer
{
    /**
    * Determine if the provided string is a valid socket address.
    *
    * This method declaration must be kept unmodified.
    *
    * @param address A string with a socket address, eg, 
    *                '127.12.23.43:5000', to be checked for validity.
    * @return boolean indicating wether the provided string is a valid
    *                  socket address
    */
    public static boolean isValidSocketAddress(String address)
    {
        Pattern pattern;
        Matcher matcher;

        String socket_address =
                        "^([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\." + //number between 0-255
                        "([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\." +  //number between 0-255
                        "([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\." +  //number between 0-255
                        "([01]?\\d\\d?|2[0-4]\\d|25[0-5]):" +    //number between 0-255
                        "(6553[0-5]|655[0-2]\\d|65[0-4]\\d{2}|6[0-4]\\d{3}|[1-5]\\d{4}|[1-9]\\d{0,3})$"; //valid por number

        pattern = Pattern.compile(socket_address);
        matcher = pattern.matcher(address);
        return matcher.matches();
    }

    /*
    * This tests your code with the examples given in the question,
    * and is provided only for your conveinience.
    */
    public static void main(String[] args)
    {
        String[] addresses = {"127.12.23.43:5000", "127.A:-12"};
        for (String address : addresses)
        {
            System.out.println(isValidSocketAddress(address));
        }
    }
}

File: python/parse_socket_address/answer.py
import re
"""
This is the file with your answer, do not rename or move it.
Write your code in it, and save it before submitting your answer.
"""

def is_valid_socket_address(socket_address):
	"""Determine if the provided string is a valid socket address.
	This function declaration must be kept unmodified.

	Args:
		socket_address: A string with a socket address, eg, 
			'127.12.23.43:5000', to be checked for validity.
	Returns:
		A boolean indicating whether the provided string is a valid 
		socket address.
	"""

	ip3Oct = "^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}" #reg ex for first 3 octetes
	ip4Oct = "(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)" #reg ex for 4 octete
	port = ":(6(?:[0-4][0-9]{3}|5(?:[0-4][0-9]{2}|5(?:[0-2][0-9]|3[0-5])))|[1-5][0-9]{4}|[1-9][0-9]{1,3}|[0-9])$" #reg exp for port number
	pat = re.compile(ip3Oct + ip4Oct + port)
	test = pat.match(socket_address)
	if test:
		return True
	else:
		return False

# This tests your code with the examples given in the question, 
# and is provided only for your convenience.
if __name__ == '__main__':
	for socket_address in ["127.12.23.43:5000",
				   "127.A:-12"]:
		print is_valid_socket_address(socket_address)


